### ISSUE TITLE
the word `a` missing in some languages

### ISSUE THREAD

**Issue by vang1ong7ang** [tatuylonen/wiktextract#267]
I downloaded the dump from <https://kaikki.org/dictionary/rawdata.html> (dated 2023-05-20).

but seems the word [`a`](https://en.wiktionary.org/wiki/a) is missing in some languages (e.g. Portuguese)

Is there any reason and is it possible to be fixed?

**Comment by kristian-clausal** [comment]
There's some kind of issue with the page in general, I'll take a look at it. Kaikki hasn't updated because it's been failing due to various bugs from big changes we've made these past couple of months, but I think it should update soon. However, there's holidays coming up so any further updates (beyond kaikki.org updating itself) might have to wait for a while.

**Comment by kristian-clausal** [comment]
[Template:quote-lite](https://en.wiktionary.org/wiki/Template:quote-lite) seems to have broken comments. Yeah, this'll be fun to figure out... :+1:

**Comment by xxyzz** [comment]
Somehow many level two nodes after the English node are included in the English node's children attribute, but they should be sibling nodes. I think this happens before expanding template quote-lite.

**Comment by kristian-clausal** [comment]
This is because of a stray `<!--` at the end of Template:quote-lite eating up a `</div>`, which eats up a ton of stuff coming after it (everything?).

**Comment by xxyzz** [comment]
You're right, `quote-lite` need to be pre-expanded and once I removed the redundant `<!--` the page was parsed correctly.

**Comment by kristian-clausal** [comment]
Removing the `<!--` on wiktionary and previewing the change with the page `a` (which is the ONLY page to use Template:quote-lite) leads to *really weird* list numbering breakdowns in the Norwegian Bokmål entry, and I have no idea why. There's also a missing `</span>`, adding it didn't help either. Thing is, the source of the HTML of the article shows that those missing tags are inserted sometime later, but I don't know how and when.

**Comment by xxyzz** [comment]
I just remove the `<!--` before the `</div>` in template "quote-lite" in sqlite db file and the final json file looks fine. Do you mean the first <span> tag in "quote-lite" is not closed? I guess your browser adds the closing tag, browsers are quite tolerant with missing tags in HTML.

**Comment by xxyzz** [comment]
I was thinking the first regex in `Wtp._template_to_body()`(`text = re.sub(r"(?s)<!\s*--.*?--\s*>", "", text)` line) should be fixed, because it removes the text between unbalanced comments. But then I found MediaWiki also discards the text after that `<!--` comment...

We could use lxml to parse the broken HTML text then convert the parse tree to text, lxml will adds the missing tags.

**Comment by kristian-clausal** [comment]
The problem is the whole structure of our parser doing things piecemeal. The HTML we work with is not complete HTML, because templates can create snippets of HTML that are continued later, and we parse those bits at different points so that we can (for example) use template arguments. To use lxml we'd have to basically just run through everything all at once, get the HTML source of a wiktionary article... And at that point, you might as well just download the HTML page of a wiktionary article and try to parse that, without any "meta"-data from templates etc.

Also, Tatu has been overzealous with all the `\s*` whitespace regex bits in HTML tags. I already removed them from the bit you quoted (`<\s*!\s*--` -> `<!--`) and I've been removing them when I see them around our codespace, and if you see something that breaks HTML's tokenizing rules please point them out or fix them.

Basically:
* HTML tags can't start with whitespace: `< tag>`, `</ tag>`, `< /tag>` are wrong and will cause misparsing if accepted. The whole entity starting with `<` and ending with the tagname is a whole unbroken token, `</?tagname`
* This includes `<nowiki>` and `<noinclude>` and comments. `<!--` and `-->` are complete tokens.
* You can have whitespace after the tagname and before the final token: `<tagname   \n>` is valid
* `/>` is a single token and can't be broken up: `<tag/ >` is invalid

**Comment by xxyzz** [comment]
I mean using lxml to add missing tags after expanding template if the expanded text has HTML tags but before calling `Wtp.parse()` with the expanded text, should be here: https://github.com/tatuylonen/wikitextprocessor/blob/ac780c16e3927bbc3570b5e4e0a295e35f54a885/wikitextprocessor/core.py#L1360

lxml will treat the wikitext between HTML tags as plain text and should just add missing tags. I guess even the expanded template is not 100% HTML but it's HTML tags should be balanced unless it's designed to be unclosed?

**Comment by kristian-clausal** [comment]
I believe there are templates that will generate HTML tags that are not meant to be closed, like the stuff going on with `{{trans-top}}` and `{{trans-bottom}}` in [Module:translations](https://en.wiktionary.org/wiki/Module:translations#L-141).

**Comment by kristian-clausal** [comment]
I have pushed a few commits that doesn't *fix* the problem, but contains it.

Basically, we already have a function `close_begline_lists()` used in parser.py that is used by token handler functions to check whether we're at the start of a new line and to close any currently open lists if we are.

However, there was a small deficiency/bug in close_begline_lists() that meant that if you had other layers of the parser-stack other than NodeKind.LISTs or .LIST_ITEMS inside of your list when you wanted to close it, it would stop the closing (popping) process there. I've changed it to check and see if the parsing stack has any NodeKind.LIST items left before it stops.

I've also sprinkled a lot more begline_close_lists() into other token handling functions that didn't have them before, like the one for sections! Now, when those token handling functions are called at the beginning of a line (meaning, they're definitely not inside a list), they will close down the list, and subsequently close down the `<div>` that was consuming everything.

This is not a fix, because the list itself is still slightly broken because it contains a div that contains the rest of the list itself, but at least this contains the damage a stray div inside a list can do to the whole rest of the document.

[ ] XXX: Do this kind of thing inside the list itself (not just outside of it). Shouldn't be too difficult, just means using something else as a check other than whether we're at the beginning of a line and a way to pop until we reach the correct depth.
[ ] XXX: Sections (`=====Noun=====` etc) should probably close all other containing entities. Basically, when a new section is started it should pop() the parser stack until it reaches the correct depth... And now that I'm looking at the code, it does that, except it obviously doesn't work well.

**Comment by xxyzz** [comment]
https://github.com/xxyzz/wikitextprocessor/commit/5df3a172aa42a68340eda6892b4deffa9dcbdd2c should fix this error but it couldn't pass some list parser tests...

**Comment by kristian-clausal** [comment]
Some of the tests are just plain wrong, and create different output compared to if you just try them out in a wikitext sandbox; I changed a couple of them when I did the changes here. If a test and actual wikitext online sandbox behavior are different, the test should be changed.

**Comment by kristian-clausal** [comment]
You're right, I'll leave it be.

I got stuck in a rabbit-hole trying to figure out why the tests
create structures that do not follow the HTML generated by wikitext.

For example, test_dl in test_parser has this:
```
 ; Mixed definition lists
 ; item 1 : definition
 :; sub-item 1 plus term
 :: two colons plus definition
 :; sub-item 2 : colon plus definition
 ; item 2
 : back to the main list
```
When you put it in a wikitext sanbox, it creates this:

```
 <dl><dt>Mixed definition lists</dt>
 <dt>item 1</dt>
 <dd>definition
 <dl><dt>sub-item 1 plus term</dt>
 <dd>two colons plus definition</dd>
 <dt>sub-item 2</dt>
 <dd>colon plus definition</dd></dl></dd></dl>
 <dl><dt>item 2</dt>
 <dd>back to the main list</dd></dl>

        DL
            DT
            DT
            DD
                DL
                    DT
                    DD
                    DT
                    DD
        DL
            DT
            DD
```

with two lists, at least, when the test asserted one child.

Turns out the tests are taken from https://www.mediawiki.org/wiki/Help:Lists#A_little_bit_further straight from help documents, and the example there explicitly says "back to main list", and this isn't something Tatu created for the test.

So, all the tests have been created with the assumption in mind that we are parsing them as they "should" be, not as what the HTML DOM eventually turns out to be...

The description list syntax (with the :; not being a "hierarchical"
string like "##" or "##:" or "*#*" is) also mean that my idea for
string comparisons wouldn't work out.

**Comment by xxyzz** [comment]
Even though MediaWiki generates two lists, the result HTML looks like a single list in a browser. Maybe some extractor code rely on the single list structure.

And the recent change to the `wikitextprocessor.parser.text_fn()` moves the text token out of the previous list item, I guess the original `text_fn()` doesn't pop the list item node to handle the typo case(extra newline in a list item text) like the `test_list_cont2` test because normally text token won't appear between list items.

**Comment by kristian-clausal** [comment]
Can you recreate the extra `\n` at the end of `quote-lite`? I only see one `\n`. What version of quote-lite do you currently have, in overrides or in the db?

**Comment by xxyzz** [comment]
I also see a single "\n" now after recreating the db from the 20230601 dump file. I just realize I added a "\n" to the quote-lite template when I edit the db file.

**Comment by kristian-clausal** [comment]
Then I think we can simplify text_fn to remove the condition. `\n\n` should (and does) break a list in wikitext into two lists, though it might be visible only in  numbered lists where the numbering gets reset.
